android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/BroadcastActivity.java

```diff
@@ -0,0 +1,207 @@
+                  Toast.makeText(this, "Couldn't find URL to broadcast", Toast.LENGTH_LONG).show();
+                }
+            }
+        }
+    }
+
+    private boolean checkLE() {
+        if (!checkIfBluetoothCanAdvertise()) {
+            Toast.makeText(this, "Bluetooth LE advertising is not supported on this device",
+                Toast.LENGTH_LONG).show();
+            finish();
+            return false;
+        }
+
+        return true;
+    }
+
+    // Check if bluetooth is on
+    private boolean checkIfBluetoothIsEnabled() {
+        BluetoothManager bluetoothManager = (BluetoothManager) getSystemService(
+            Context.BLUETOOTH_SERVICE);
+        BluetoothAdapter bluetoothAdapter = bluetoothManager.getAdapter();
+        if (bluetoothAdapter == null || !bluetoothAdapter.isEnabled()) {
+          Log.d(TAG, "not enabled");
+            return false;
+        }
+        Log.d(TAG, "enabled");
+        return true;
+    }
+
+    // Ask the user to enable bluetooth
+    private void requestBluetooth() {
+        Intent enableBtIntent = new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE);
+        startActivityForResult(enableBtIntent, ENABLE_BLUETOOTH_REQUEST_ID);
+    }
+
+    // Check if the given bluetooth hardware
+    // on the current device supports ble advertisements
+    @TargetApi(21)
+    private boolean checkIfBluetoothCanAdvertise() {
+        BluetoothManager bluetoothManager = (BluetoothManager) getSystemService(
+            Context.BLUETOOTH_SERVICE);
+        if (bluetoothManager.getAdapter().getBluetoothLeAdvertiser() == null) {
+            Log.d(TAG, "cant advertise");
+            return false;
+        }
+        Log.d(TAG, "can advertise");
+        return true;
+    }
+
+    private void startBroadcastService(String url){
+        Intent inten = new Intent(this, PhysicalWebBroadcastService.class);
+        inten.putExtra("URL", url);
+        inten.putExtra("OURL", displayUrl);
+        startService(inten);
+    }
+
+
+    // Link up to the service so that this activity
+    // can call the service's public methods
+    // Check if the given text contains a URL
+    private String findUrlInText(String text) {
+        List<String> urls = new ArrayList<>();
+        Matcher m = Patterns.WEB_URL.matcher(text);
+        while (m.find()) {
+            String url = m.group();
+            urls.add(url);
+        }
+        if (urls.size() > 0) {
+            return urls.get(0);
+        }
+
+        return null;
+    }
+
+    private static boolean hasValidUrlLength(String url) {
+        int uriLength = UriBeacon.uriLength(url);
+        return 0 <= uriLength && uriLength <= MAX_URI_LENGTH;
+    }
+    private static boolean isAsciiUrl(String url) {
+        boolean isCompliant = false;
+        try {
+          URI uri = new URI(url);
+          String urlString = uri.toASCIIString();
+          isCompliant = url.equals(urlString);
+        } catch (URISyntaxException e) {
+          // bad url
```

Comment:
What do we do here?  toast andfinish "Sorry we do not yet support sharing non-ascii URLs"